doc/indexing.rst

```diff
@@ -353,67 +331,129 @@ labels of the 1st and 2nd dimension, respectively. You must specify all
+
+
+Furthermore, you can use multi-dimensional :py:meth:`~xarray.DataArray`
+as indexers, where the resultant array dimension is also determined by
+indexers' dimension,
+
+.. ipython:: python
+
+    ind = xr.DataArray([[0, 1], [0, 1]], dims=['a', 'b'])
+    da[ind]
+
+To summarize, our advanced indexing is based on our broadcasting scheme.
+See :ref:`compute.broadcasting` for the detail.
 
-Orthogonal (outer) vs. vectorized indexing
-------------------------------------------
 
-Indexing with xarray objects has one important difference from indexing numpy
-arrays: you can only use one-dimensional arrays to index xarray objects, and
-each indexer is applied "orthogonally" along independent axes, instead of
-using numpy's broadcasting rules to vectorize indexers. This means you can do
-indexing like this, which would require slightly more awkward syntax with
-numpy arrays:
+These advanced indexing also works with ``isel``, ``loc``, and ``sel``.
 
 .. ipython:: python
 
-    arr[arr['time.day'] > 1, arr['space'] != 'IL']
+    ind = xr.DataArray([[0, 1], [0, 1]], dims=['a', 'b'])
+    da.isel(y=ind)  # same to da[:, ind]
+
+    ind = xr.DataArray([['a', 'b'], ['b', 'a']], dims=['a', 'b'])
+    da.loc[:, ind]  # same to da.sel(y=ind)
+
+
+and also for Dataset
+
+.. ipython:: python
+
+    ds2 = da.to_dataset(name='bar')
+    ds2.isel(x=xr.DataArray([0, 1, 2], dims=['points']))
+
+.. note::
+  This advanced indexing was newly added in v.0.10.
+  In the older version of xarray, dimensions of indexers are not used.
+  Special methods to realize some advanced indexing,
+  ``isel_points`` and ``sel_points`` are now deprecated.
 
-This is a much simpler model than numpy's `advanced indexing`__. If you would
-like to do advanced-style array indexing in xarray, you have several options:
 
-* :ref:`pointwise indexing`
-* :ref:`masking with where`
-* Index the underlying NumPy array directly using ``.values``, e.g.,
+More advanced indexing
+-----------------------
 
-__ http://docs.scipy.org/doc/numpy/reference/arrays.indexing.html
+The use of :py:meth:`~xarray.DataArray` as indexers enables very flexible indexing.
+The following is an example of the pointwise indexing,
 
 .. ipython:: python
 
-    arr.values[arr.values > 0.5]
+    # index by integer array indices
+    da = xr.DataArray(np.arange(56).reshape((7, 8)), dims=['x', 'y'])
+    da
+    da.isel(x=xr.DataArray([0, 1, 6], dims='z'),
+            y=xr.DataArray([0, 1, 0], dims='z'))
+
+where three elements at ``(ix, iy) = ((0, 0), (1, 1), (6, 0))`` are selected
+and mapped along a new dimension ``z``.
+
+If you want to add a coordinate to the new dimension ``z``,
+you can supply a :py:meth:`~xarray.DataArray` with a coordinate,
+
+.. ipython:: python
+
+    da.isel(x=xr.DataArray([0, 1, 6], dims='z',
+                           coords={'z': ['a', 'b', 'c']}),
+            y=xr.DataArray([0, 1, 0], dims='z'))
+
+
+Analogously, label-based pointwise-indexing is also possible by ``.sel`` method,
```

Comment:
It would be really nice to also have a power-user example of pointwise indexing with 2D indexers and nearest-neighbor lookups, e.g., to switch to another coordinate projection. Something like `ds.sel(latitude=latitude_grid, longitude=longitude_grid, method='nearest', tolerance=0.1)`.

But that should probably go on another page.